cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -536,6 +545,311 @@ aggregate_reader_metadata::aggregate_reader_metadata(
       CUDF_EXPECTS(schema == pfm.schema, "All sources must have the same schema");
     }
   }
+
+  if (arrow_schema.has_value()) {
+    // consume arrow schema into the parquet schema
+    consume_arrow_schema();
+    // we no longer keep arrow schema alive
+    arrow_schema.reset();
+  }
+
+  // erase "ARROW:schema" from the output pfm if exists
+  std::for_each(
+    keyval_maps.begin(), keyval_maps.end(), [](auto& pfm) { pfm.erase("ARROW:schema"); });
+}
+
+[[nodiscard]] std::optional<arrow_schema_data_types>
+aggregate_reader_metadata::collect_arrow_schema(bool use_arrow_schema) const
+{
+  // Check if we want to collect the arrow_schema
+  if (not use_arrow_schema) { return std::nullopt; }
+
+  // Check the key_value metadata for ARROW:schema, decode and walk it
+  // Function to convert from flatbuf::duration type to cudf::type_id
+  auto duration_from_flatbuffer = [](flatbuf::Duration const* duration) {
```

Comment:
```suggestion
  auto const duration_from_flatbuffer = [](flatbuf::Duration const* duration) {
```
nit: const when possible